### PR TITLE
Phase2-hgx84 Add functionality to HGCal geometry to access information

### PR DESCRIPTION
**Pull request by bsunanda**
Regarding wafer numbers in a given layer or z position of each layer

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: Geometry/HGCalCommonData/interface/HGCalDDDConstants.h [hunk 1]
```diff
@@ -102,13 +101,16 @@ class HGCalDDDConstants {
...
   bool waferInLayer(int wafer, int lay) const;
 
-  const HGCalParameters* hgpar_;
-  constexpr static double tan30deg_ = 0.5773502693;
-  double                 rmax_, hexside_;
-  HGCalGeometryMode      mode_;
-  int32_t                tot_wafers_, modHalf_;
-  std::array<uint32_t,2> tot_layers_;
-  simrecovecs            max_modules_layer_; 
+  typedef std::array<std::vector<int32_t>, 2> simrecovecs;
```
Comment (ianna) [inline Geometry/HGCalCommonData/interface/HGCalDDDConstants.h]: @bsunanda - please, use capitalised name for typedef

File: Geometry/HGCalCommonData/interface/HGCalDDDConstants.h [hunk 2]
```diff
@@ -102,13 +101,16 @@ class HGCalDDDConstants {
 			  float& tl, float& alpha) const;
   bool waferInLayer(int wafer, int lay) const;
```
Comment (ianna) [inline Geometry/HGCalCommonData/interface/HGCalDDDConstants.h]: @bsunanda - please, check that calculations are necessary in the member function. For example, line 806 seems to be equal to hexside_  = 2.0 * rmax_ * tan30deg_;

**Comment by ianna** [comment]
@bsunanda - please, check that this 54 package dependency on these 4 Geometry packages is justified:
```
Merge made by the 'recursive' strategy.
 Geometry/CaloGeometry/src/FlatTrd.cc                          |   6 +++-
 Geometry/HGCalCommonData/interface/HGCalDDDConstants.h        |  20 ++++++-----
 Geometry/HGCalCommonData/src/HGCalDDDConstants.cc             |  45 ++++++++++++++++++-----
 Geometry/HGCalGeometry/test/BuildFile.xml                     |   3 ++
 Geometry/HGCalGeometry/test/HGCalTestRecHitTool.cc            | 180 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 Geometry/HGCalGeometry/test/python/testHGCalRecHitTool_cfg.py |  50 ++++++++++++++++++++++++++
 Geometry/HcalCommonData/interface/HcalDDDRecConstants.h       |   1 +
 Geometry/HcalCommonData/src/HcalDDDRecConstants.cc            |  14 ++++++++
 8 files changed, 301 insertions(+), 18 deletions(-)
 create mode 100644 Geometry/HGCalGeometry/test/HGCalTestRecHitTool.cc
 create mode 100644 Geometry/HGCalGeometry/test/python/testHGCalRecHitTool_cfg.py
Switched to branch 'from-CMSSW_9_2_X_2017-06-08-2300'
>> Checking Geometry/CaloGeometry CMSSW_9_2_X_2017-06-08-2300
   x Geometry/CaloGeometry/src/FlatTrd.cc
>> Checking Geometry/HGCalCommonData CMSSW_9_2_X_2017-06-08-2300
   x Geometry/HGCalCommonData/interface/HGCalDDDConstants.h
   x Geometry/HGCalCommonData/src/HGCalDDDConstants.cc
>> Checking Geometry/HGCalGeometry CMSSW_9_2_X_2017-06-08-2300
   x Geometry/HGCalGeometry/test/BuildFile.xml
>> Checking Geometry/HcalCommonData CMSSW_9_2_X_2017-06-08-2300
   x Geometry/HcalCommonData/interface/HcalDDDRecConstants.h
   x Geometry/HcalCommonData/src/HcalDDDRecConstants.cc
Checking out these packages: 54
CalibCalorimetry/CaloMiscalibTools (header)
CalibCalorimetry/CaloTPG (header)
CalibCalorimetry/CastorCalib (header)
CalibCalorimetry/HcalAlgos (header)
CalibCalorimetry/HcalPlugins (header)
CalibCalorimetry/HcalTPGAlgos (header)
CalibCalorimetry/HcalTPGEventSetup (header)
Calibration/HcalAlCaRecoProducers (header)
Calibration/HcalCalibAlgos (header)
Calibration/IsolatedParticles (header)
CaloOnlineTools/HcalOnlineDb (header)
CondFormats/HcalObjects (header)
CondTools/Geometry (header)
CondTools/Hcal (header)
DQMOffline/Hcal (header)
DQMOffline/JetMET (header)
EventFilter/CastorRawToDigi (header)
Fireworks/Geometry (header)
GeneratorInterface/RivetInterface (header)
Geometry/CaloEventSetup (header)
Geometry/CaloTopology (header)
Geometry/HcalEventSetup (header)
Geometry/HcalTowerAlgo (header)
HLTrigger/JetMET (header)
L1Trigger/L1TCaloLayer1 (header)
L1Trigger/L1THGCal (header)
L1TriggerConfig/L1ScalesProducers (header)
RecoJets/FFTJetProducers (header)
RecoJets/JetProducers (header)
RecoLocalCalo/CaloTowersCreator (header)
RecoLocalCalo/HGCalRecAlgos (header)
RecoLocalCalo/HGCalRecProducers (header)
RecoLocalCalo/HcalRecAlgos (header)
RecoLocalCalo/HcalRecProducers (header)
RecoMET/METProducers (header)
RecoParticleFlow/PFClusterProducer (header)
RecoParticleFlow/PFTracking (header)
RecoTauTag/RecoTau (header)
SimCalorimetry/HGCalSimProducers (header)
SimCalorimetry/HcalSimAlgos (header)
SimCalorimetry/HcalSimProducers (header)
SimCalorimetry/HcalTrigPrimAlgos (header)
SimCalorimetry/HcalTrigPrimProducers (header)
SimG4CMS/Calo (header)
SimG4CMS/HGCalTestBeam (header)
SimGeneral/CaloAnalysis (header)
SimGeneral/DataMixingModule (header)
TopQuarkAnalysis/TopEventProducers (header)
Validation/GlobalHits (header)
Validation/GlobalRecHits (header)
Validation/HGCalValidation (header)
Validation/HcalDigis (header)
Validation/HcalHits (header)
Validation/HcalRecHits (header)
[yana@cmsdev02 src]$ ls
CalibCalorimetry  CondFormats  EventFilter         Geometry   L1TriggerConfig  RecoMET           SimCalorimetry  TopQuarkAnalysis
Calibration       CondTools    Fireworks           HLTrigger  RecoJets         RecoParticleFlow  SimG4CMS        Validation
CaloOnlineTools   DQMOffline   GeneratorInterface  L1Trigger  RecoLocalCalo    RecoTauTag        SimGeneral

```

**Comment by ianna** [comment]
@bsunanda - rebase is needed

**Comment by ianna** [comment]
@bsunanda and @kpedro88 - Just a reminder, the Hcal digitizer geometry check shows rather high on the performance profiling list and should be unnecessary: EventSetup takes care of its record updates and dependency propagation when it is declared.
```
   12.4      23.43  HcalDigitizer::checkGeometry(edm::EventSetup const&) [45]
```
http://cmslxr.fnal.gov/source/SimCalorimetry/HcalSimProducers/src/HcalDigitizer.cc#0551

**Comment by kpedro88** [comment]
@ianna `HcalDigitizer::updateGeometry()` does a number of things beyond basic EventSetup retrieval. We can try to see if this can be called less often.

File: Geometry/HGCalCommonData/interface/HGCalDDDConstants.h (hunk 2 above)
Comment (bsunanda) [inline Geometry/HGCalCommonData/interface/HGCalDDDConstants.h]: I shall check this for a later version. This method was not touched in this change

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by ianna** [comment]
+1

File: Geometry/HGCalCommonData/src/HGCalDDDConstants.cc [hunk 3]
```diff
@@ -713,6 +730,18 @@ int HGCalDDDConstants::wafers() const {
   return wafer;
 }
 
+int HGCalDDDConstants::wafers(int layer, int type) const {
+
+  int  wafer(0);
+  auto itr = waferLayer_.find(layer);
+  if (itr != waferLayer_.end()) {
+    if      (type == 2) wafer = std::get<2>(itr->second);
+    else if (type == 2) wafer = std::get<1>(itr->second);
```
Comment (kpedro88) [inline Geometry/HGCalCommonData/src/HGCalDDDConstants.cc]: looks like a typo - same condition `type==2` repeated twice

File: Geometry/HGCalGeometry/test/HGCalTestRecHitTool.cc [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+  }
+}
+
+std::vector<double> HGCalTestRecHitTool::retrieveLayerPositions(unsigned layers) {
+
+  DetId id;
+  std::vector<double> layerPositions;
+  for (unsigned ilayer=1; ilayer<=layers; ++ilayer) {
+    int lay(ilayer), type(0);
+    if (ilayer<=28) {
```
Comment (kpedro88) [inline Geometry/HGCalGeometry/test/HGCalTestRecHitTool.cc]: @bsunanda we really need a way to get these magic numbers from a geometry/topology class

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: Geometry/HGCalGeometry/test/HGCalTestRecHitTool.cc (hunk 4 above)
Comment (bsunanda) [inline Geometry/HGCalGeometry/test/HGCalTestRecHitTool.cc]: @kpedro88 I shall look into this

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@ianna @kpedro88 Could you approve this?

**Comment by ianna** [comment]
+1

**Comment by kpedro88** [comment]
+1

**Comment by kpedro88** [comment]
@clelange please see ae1abb5 for an example of how to get magic numbers from the HGCal geometry - this should be propagated to the reco algorithms

**Comment by clelange** [comment]
@kpedro88 OK, great, I'll try to address that towards the end of the week/next week.

File: Geometry/HGCalCommonData/interface/HGCalDDDConstants.h [hunk 5]
```diff
@@ -102,13 +101,16 @@ class HGCalDDDConstants {
...
 
+  typedef std::array<std::vector<int32_t>, 2> simrecovecs;
```
Comment (davidlange6) [inline Geometry/HGCalCommonData/interface/HGCalDDDConstants.h]: please use an upper case letter to start type names - eg, SimRecoVecs

File: Geometry/HGCalCommonData/interface/HGCalDDDConstants.h [hunk 6]
```diff
@@ -102,13 +101,16 @@ class HGCalDDDConstants {
 			  float& tl, float& alpha) const;
   bool waferInLayer(int wafer, int lay) const;
 
+  typedef std::array<std::vector<int32_t>, 2> simrecovecs;  
+  typedef std::tuple<int,int,int> HGCWaferParam;
```
Comment (davidlange6) [inline Geometry/HGCalCommonData/interface/HGCalDDDConstants.h]: make this a std::array<int,3> and lots of code gets simpler

File: Geometry/HGCalCommonData/src/HGCalDDDConstants.cc [hunk 7]
```diff
@@ -713,6 +730,18 @@ int HGCalDDDConstants::wafers() const {
   return wafer;
 }
 
+int HGCalDDDConstants::wafers(int layer, int type) const {
+
+  int  wafer(0);
+  auto itr = waferLayer_.find(layer);
+  if (itr != waferLayer_.end()) {
+    if      (type == 2) wafer = std::get<2>(itr->second);
```
Comment (davidlange6) [inline Geometry/HGCalCommonData/src/HGCalDDDConstants.cc]: eg, all of this becomes wafer=itr->second[type]

File: Geometry/HGCalGeometry/test/HGCalTestRecHitTool.cc [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+
+  if (pG.isValid()) {
+    geom_ = pG.product();
+    auto geomEE = static_cast<const HGCalGeometry*>(geom_->getSubdetectorGeometry(DetId::Forward,ForwardSubdetector::HGCEE));
+    layerEE_    = (geomEE->topology().dddConstants()).layers(true);
+    auto geomFH = static_cast<const HGCalGeometry*>(geom_->getSubdetectorGeometry(DetId::Forward,ForwardSubdetector::HGCHEF));
+    layerFH_    = (geomFH->topology().dddConstants()).layers(true);
+    auto geomBH = static_cast<const HcalGeometry*>(geom_->getSubdetectorGeometry(DetId::Hcal,HcalSubdetector::HcalEndcap));
+    layerBH_    = (geomBH->topology().dddConstants())->getMaxDepth(1);
+    std::cout << "Layers " << layerEE_ << ":" << layerFH_ << ":" << layerBH_
```
Comment (davidlange6) [inline Geometry/HGCalGeometry/test/HGCalTestRecHitTool.cc]: message logger...

File: Geometry/HGCalGeometry/test/HGCalTestRecHitTool.cc [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+  if (nullptr == ddd) {
+    throw cms::Exception("HGCalTestRecHitTools")
+      << "DDDConstants not accessibl to HGCalTestRecHitTools!";
+  }
+}
+
+std::vector<double> HGCalTestRecHitTool::retrieveLayerPositions(unsigned layers) {
+
+  DetId id;
+  std::vector<double> layerPositions;
```
Comment (davidlange6) [inline Geometry/HGCalGeometry/test/HGCalTestRecHitTool.cc]: reserve  the needed length

**Comment by clelange** [comment]
I realised that there is a lot of duplication from https://github.com/cms-sw/cmssw/blob/master/RecoLocalCalo/HGCalRecAlgos/src/RecHitTools.cc in https://github.com/cms-sw/cmssw/pull/19147/files/ae1abb55ffe5e4a4c1bb8c0a39df219b21192737#diff-99ed6da389cb3ffe7a69878ebdb2892d, but the latter is defined as a plugin and named "HGCalTestRecHitTool.cc". Is this intended? And should `HGCalRecAlgos/src/RecHitTools.cc` then basically be stripped by these functions?

**Comment by bsunanda** [comment]
@clelange I was testing some of the functionalities - once tested I shall put in some of tested codes in RecHitTools.cc
@davidlange6 This is a test function and the tested code will go to RecHitTools.cc

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by kpedro88** [comment]
+1

**Comment by ianna** [comment]
+1

**Comment by bsunanda** [comment]
@davidlange6 Could you pllease merge this PR

File: Geometry/HGCalCommonData/src/HGCalDDDConstants.cc [hunk 10]
```diff
@@ -713,6 +731,18 @@ int HGCalDDDConstants::wafers() const {
   return wafer;
 }
 
+int HGCalDDDConstants::wafers(int layer, int type) const {
+
+  int  wafer(0);
+  auto itr = waferLayer_.find(layer);
+  if (itr != waferLayer_.end()) {
```
Comment (davidlange6) [inline Geometry/HGCalCommonData/src/HGCalDDDConstants.cc]: hi @bsunanda - could this not be simplified as I had suggested?

Comment (bsunanda) [inline Geometry/HGCalCommonData/src/HGCalDDDConstants.cc]: What did you suggest for this?

Comment (bsunanda) [inline Geometry/HGCalCommonData/src/HGCalDDDConstants.cc]: What did you suggest for this? I am using the array as you suggested

Comment (davidlange6) [inline Geometry/HGCalCommonData/src/HGCalDDDConstants.cc]: eg

wafer = (itr->second)[type)
?

Comment (bsunanda) [inline Geometry/HGCalCommonData/src/HGCalDDDConstants.cc]: It was not too far from what you wrote. I am doing what you have mentioned.

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@ianna @kpedro88 Please approve this

File: Geometry/HGCalCommonData/src/HGCalDDDConstants.cc (hunk 10 above)
Comment (davidlange6) [inline Geometry/HGCalCommonData/src/HGCalDDDConstants.cc]: I'm not following - you are doing
if ( type==2 ) wafer=(itr->second)[2]
if ( type==1 ) wafer=(itr->second)[1]
etc..

**Comment by ianna** [comment]
+1

**Comment by kpedro88** [comment]
+1

File: Geometry/HGCalCommonData/src/HGCalDDDConstants.cc (hunk 10 above)
Comment (bsunanda) [inline Geometry/HGCalCommonData/src/HGCalDDDConstants.cc]: Could you check the latest changes and approve this for integration

**Comment by bsunanda** [comment]
@davidlange6 Please accept it

**Comment by kpedro88** [comment]
@bsunanda needs rebase...

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by davidlange6** [comment]
merge

**Pull request merged**